Make RepairTokenRangeSplitter the default implementation

tolbertam opened the pull request:
Summary of impacting changes:

* RepairTokenRangeSplitter is now the default in favor of FixedSplitTokenRangeSplitter.

* number_of_subranges moved from a repair_type_override/global config to a property of FixedSplitTokenRangeSplitter.

* [get|set]autorepairconfig usability changes

Detailed breakdown of changes:

0. IAutoRepairTokenRangeSplitter changes:

  * Move RepairType from an parameter in getRepairAssignments to a parameter in the constructor for implementations.  This was done because the RepairType is always the same for a splitter instance.

  * Add setParameter and getParameter methods which are used by setautorepairconfig and getautorepairconfig to dynamically update splitter configuration.

1. [get|set]autorepairconfig changes:

  * getautorepairconfig output now shows property names instead of human readable names (e.g. repair_check_interval instead of repair eligibility check interval). This was done to make it more intuitive to know what properties to use for setautorepairconfig.

  * getautorepairconfig and setautorepairconfig now support viewing and changing splitter properties, e.g.:

    setautorepairconfig token_range_splitter.max_bytes_per_schedule 500GiB -t full

2. RepairTokenRangeSplitter changes:

  * Renames RepairRangeSplitter to RepairTokenRangeSplitter and makes it the default implementation.

  * Establishes defaults for each repair type to be sensible.

  * Improve javadocs detailing the primary goal of the splitter, its configuration and its defaults and the justifications for using them.

  * Rename variables to be consistent with their setting names.

3. FixedSplitTokenRangeSplitter changes:

  * Renames DefaultAutoRepairTokenSplitter to FixedSplitTokenRangeSplitter as it is no longer the default.

  * Move number_of_subranges from a global config to a property for this splitter.

4. RepairAssignmentIterator

  * Refactored common code from both splitter implementations into RepairAssignmentIterator with the aim to reduce the amount of boiler plate custom splitter implementations need to implement.

5. Test changes

  * Fix AutoRepairParameterizedTest to use fixed splitter so we get a deterministic repair plan.

  * Allow splitter to be changed programmatically, only expect it to be used for tests.

  * Rename CassandraSreamReceiverTest and fix it

    Whether streaming cdc/mvs into commitlogs was previously dependent on system properties;  update the test to account for the new yaml properties.

  * Fix dtest after CASSANDRA-20160

    The introduction of repair_task_min_duration causes repairs to take cumulatively longer for a node than 2 minutes.  To resolve this, set that to 0s, and also enable repair_by_keyspace and set subranges to 1 to reduce the overall number of repairs.

Patch by Andy Tolbert; reviewed by ___ for CASSANDRA-20179

tolbertam committed:
Make RepairTokenRangeSplitter the default implementation

Summary of impacting changes:

* RepairTokenRangeSplitter is now the default in favor of
  FixedSplitTokenRangeSplitter.

* number_of_subranges moved from a repair_type_override/global config
  to a property of FixedSplitTokenRangeSplitter.

* [get|set]autorepairconfig usability changes

Detailed breakdown of changes:

0. IAutoRepairTokenRangeSplitter changes:

  * Move RepairType from an parameter in getRepairAssignments to
    a parameter in the constructor for implementations.  This was done
    because the RepairType is always the same for a splitter instance.

  * Add setParameter and getParameter methods which are used by
    setautorepairconfig and getautorepairconfig to dynamically update
    splitter configuration.

1. [get|set]autorepairconfig changes:

  * getautorepairconfig output now shows property names instead of
    human readable names (e.g. repair_check_interval instead of
    repair eligibility check interval). This was done to make it more
    intuitive to know what properties to use for setautorepairconfig.

  * getautorepairconfig and setautorepairconfig now support viewing
    and changing splitter properties, e.g.:

    setautorepairconfig token_range_splitter.max_bytes_per_schedule 500GiB -t full

2. RepairTokenRangeSplitter changes:

  * Renames RepairRangeSplitter to RepairTokenRangeSplitter and makes it
    the default implementation.

  * Establishes defaults for each repair type to be sensible.

  * Improve javadocs detailing the primary goal of the splitter, its
    configuration and its defaults and the justifications for using them.

  * Rename variables to be consistent with their setting names.

3. FixedSplitTokenRangeSplitter changes:

  * Renames DefaultAutoRepairTokenSplitter to
    FixedSplitTokenRangeSplitter as it is no longer the default.

  * Move number_of_subranges from a global config to a property for
    this splitter.

4. RepairAssignmentIterator

  * Refactored common code from both splitter implementations into
    RepairAssignmentIterator with the aim to reduce the amount of
    boiler plate custom splitter implementations need to implement.

5. Test changes

  * Fix AutoRepairParameterizedTest to use fixed splitter so we get a
    deterministic repair plan.

  * Allow splitter to be changed programmatically, only expect it to be
    used for tests.

  * Rename CassandraSreamReceiverTest and fix it

    Whether streaming cdc/mvs into commitlogs was previously dependent
    on system properties;  update the test to account for the new
    yaml properties.

  * Fix dtest after CASSANDRA-20160

    The introduction of repair_task_min_duration causes repairs to take
    cumulatively longer for a node than 2 minutes.  To resolve this,
    set that to 0s, and also enable repair_by_keyspace and set subranges
    to 1 to reduce the overall number of repairs.

Patch by Andy Tolbert; reviewed by ___ for CASSANDRA-20179

tolbertam committed:
JDK 8 support and test rename

jaydeepkumar1984 commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepair.java:
```diff
@@ -107,7 +107,43 @@ protected AutoRepair()
+    }
+
+    @VisibleForTesting
+    static IAutoRepairTokenRangeSplitter newAutoRepairTokenRangeSplitter(AutoRepairConfig.RepairType repairType, ParameterizedClass parameterizedClass) throws ConfigurationException
```
Isn't this API (_newAutoRepairTokenRangeSplitter_) already declared and defined inside AutoRepairConfig.java?

tolbertam replied:
good catch, fixed!

tolbertam committed:
Fix newAutoRepairTokenRangeSplitter duplication

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java:
```diff
@@ -308,6 +307,26 @@ public ParameterizedClass getTokenRangeSplitter(RepairType repairType)
+    /**
+     * Set a new token range splitter, this is not meant to be used other than for testing.
+     */
+    public void setTokenRangeSplitter(RepairType repairType, ParameterizedClass tokenRangeSplitter)
```
maybe annotate with `@VisibleForTesting` , and can be package private
```suggestion
   @VisibleForTesting
   void setTokenRangeSplitter(RepairType repairType, ParameterizedClass tokenRangeSplitter)
```

tolbertam replied:
agreed, i was toying with the idea of being able to dynamically replace the splitter, but don't think its worth supporting as it may be a bit unpredictable.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java:
```diff
@@ -308,6 +307,26 @@ public ParameterizedClass getTokenRangeSplitter(RepairType repairType)
+
+    public IAutoRepairTokenRangeSplitter getTokenRangeSplitterInstance(RepairType repairType)
+    {
+        IAutoRepairTokenRangeSplitter splitter = tokenRangeSplitters.get(repairType);
+        if (splitter == null)
+        {
+            splitter = newAutoRepairTokenRangeSplitter(repairType, getTokenRangeSplitter(repairType));
+            tokenRangeSplitters.put(repairType, splitter);
+        }
+        return splitter;
```
Prefer `computeIfAbsent`
```suggestion
        return tokenRangeSplitters.computeIfAbsent(repairType,
                                                   key -> newAutoRepairTokenRangeSplitter(key, getTokenRangeSplitter(key)));
```

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairConfig.java:
```diff
@@ -328,6 +347,43 @@ public void setRepairSessionTimeout(RepairType repairType, String repairSessionT
+    {
+        try
+        {
+            Class<?> tokenRangeSplitterClass;
```
nit: to avoid casting below
```suggestion
            Class<? extends IAutoRepairTokenRangeSplitter> tokenRangeSplitterClass;
```

frankgh reviewed (commented):
This is a WIP review, I will continue later

tolbertam commented on src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java:
```diff
@@ -61,163 +63,258 @@
+ *         <b>full</b>:  Configured in a way that attempts to accomplish repairing all data in a schedule, with
+ *         individual repairs targeting at most 200GiB of data and 1048576 partitions.
+ *         <ul>
+ *             <li><b>bytes_per_assignment</b>: 200GiB</li>
```
Thoughts on this?  My original rationale was that the `partitions_per_assignment` guardrail would do well to prevent over-streaming, but allowing ~200GB validations might be a bit much.

Although, this doesn't factor in compression, so if you have ~2TB nodes and 50% compression, it'd likely create 20 splits, so seems about right.

Intending to try this out in practice this weekend.

tolbertam committed:
Nit-fixing in existing code

* Move curly brackets to new line
* Remove unused config declaration in constructor

tolbertam committed:
Process feedback

Also remove redundant AutoRepairConfig.RepairType for RepairType
in AutoRepairConfig.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/AutoRepairUtils.java:
```diff
@@ -430,7 +430,7 @@ protected static TreeSet<UUID> getHostIdsInCurrentRing(RepairType repairType, Se
             else
             {
-                logger.info("Node is not present in Gossipe cache node {}, node data center {}", node, nodeDC);
+                logger.info("Node is not present in Gossip cache node {}, node data center {}", node, nodeDC);
```
unrelated to your PR, but should this be `warn`?

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -33,68 +34,49 @@
 {
-    @Override
-    public Iterator<KeyspaceRepairAssignments> getRepairAssignments(AutoRepairConfig.RepairType repairType, boolean primaryRangeOnly, List<PrioritizedRepairPlan> repairPlans)
+    /**
+     * The number of subranges to split each to-be-repaired token range into,
+     * the higher this number, the smaller the repair sessions will be
+     * How many subranges to divide one range into? The default is 1.
+     * If you are using v-node, say 256, then the repair will always go one v-node range at a time, this parameter, additionally, will let us further subdivide a given v-node range into sub-ranges.
+     * With the value “1” and v-nodes of 256, a given table on a node will undergo the repair 256 times. But with a value “2,” the same table on a node will undergo a repair 512 times because every v-node range will be further divided by two.
+     * If you do not use v-nodes or the number of v-nodes is pretty small, say 8, setting this value to a higher number, say 16, will be useful to repair on a smaller range, and the chance of succeeding is higher.
+     */
```
Small suggestion for better readability. Also, this documentation feels like it belongs in cassandra.yaml and not hidden in the code.

```suggestion
        /**
     * The number of subranges to split each to-be-repaired token range into.
     * The higher this number, the smaller the repair sessions will be
     * How many subranges to divide one range into? The default is 1.
     * If you are using v-nodes, say 256, then the repair will always go one
     * v-node range at a time, this parameter, additionally, will let us
     * further subdivide a given v-node range into sub-ranges.
     * With the value "1" and v-nodes of 256, a given table on a node will
     * undergo the repair 256 times. But with a value "2", the same table
     * on a node will undergo a repair 512 times because every v-node range
     * will be further divided by two.
     * If you do not use v-nodes or the number of v-nodes is pretty small,
     * say 8, setting this value to a higher number, say 16, will be useful
     * to repair on a smaller range, and the chance of succeeding is higher.
     */
```

tolbertam replied:
Agreed, I ported this over from `AutoRepairConfig` to the splitter itself as is.   I think we need to make dedicated docs on this and also notably update cassandra.yaml as you mention.

I think we need to take a pass at this generally, as you can see in `RepairTokenRangeSplitter` I added a bunch of documentation around defaults, and `AutoRepairConfig` itself has a bunch.  Will implement your suggestion, but mind if we tackle this as another PR?

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -33,68 +34,49 @@
-        private PrioritizedRepairPlan currentPlan = null;
-
-        RepairAssignmentIterator(AutoRepairConfig.RepairType repairType, boolean primaryRangeOnly, List<PrioritizedRepairPlan> repairPlans)
+        if (parameters.containsKey(NUMBER_OF_SUBRANGES))
```
minor nit:
```suggestion
        this.numberOfSubranges = Integer.parseInt(parameters.getOrDefault(NUMBER_OF_SUBRANGES, "1"));
```

tolbertam replied:
:+1: will do the same thing in `RepairTokenRangeSplitter`

tolbertam replied:
also notable, the default here should be 16, as thats what it was in `AutoRepairConfig`, i'll revert it back (and also update the comment)

tolbertam replied:
actually, i'll just keep at one given the comments about vnodes.  Since the splitter needs to be enabled explicitly now (since it's no longer the default), it'll be up to the implementor in terms of what to set it to.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -139,4 +120,34 @@ private KeyspaceRepairAssignments getRepairAssignmentsForKeyspace(AutoRepairConf
+    @Override
+    public Map<String, String> getParameters()
+    {
+        return Collections.unmodifiableMap(new LinkedHashMap<String, String>()
+        {{
+            put(NUMBER_OF_SUBRANGES, Integer.toString(getNumberOfSubranges()));
+        }});
```
we can just use a singleton map, which is already immutable :
```suggestion
        return Collections.singletonMap(NUMBER_OF_SUBRANGES, Integer.toString(getNumberOfSubranges()));
```

tolbertam replied:
so much simpler, didn't realize that was available in JDK8,  will apply this in other places I used this as well.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -139,4 +120,34 @@ private KeyspaceRepairAssignments getRepairAssignmentsForKeyspace(AutoRepairConf
+        if (key.equals(NUMBER_OF_SUBRANGES))
+        {
+            setNumberOfSubranges(Integer.parseInt(value));
+        }
```
no return here? this will fall through and throw always. maybe invert the if statement `if (!key....) throw...`

tolbertam replied:
oh right, that's a clear bug.  Thanks for pointing this out.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/FixedSplitTokenRangeSplitter.java:
```diff
@@ -107,7 +89,6 @@ private KeyspaceRepairAssignments getRepairAssignmentsForKeyspace(AutoRepairConf
             // if we need to repair non-primary token ranges, then change the tokens accrodingly
```
```suggestion
            // if we need to repair non-primary token ranges, then change the tokens accordingly
```

jaydeepkumar1984 commented on src/java/org/apache/cassandra/repair/autorepair/RepairTokenRangeSplitter.java:
```diff
@@ -427,9 +524,9 @@ static SizedRepairAssignment merge(List<SizedRepairAssignment> assignments)
+        List<SizeEstimate> sizeEstimates = getRangeSizeEstimate(keyspaceName, tableName, tokenRanges);
         return getRepairAssignments(sizeEstimates);
     }
 
```
cfs null check at ?
`ColumnFamilyStore cfs = ColumnFamilyStore.getIfExists(estimate.keyspace, estimate.table);
                long memtableSize = cfs.getTracker().getView().getCurrentMemtable().getLiveDataSize();`

tolbertam replied:
good call, i ended up amending this to skip over tables that don't have a ColumnFamilyStore and logging a warning instead.  I'm betting this would be encounter-able when a table is dropped, in this case it seems reasonable to skip over.

I also changed the code such that we always return a repair assignment even if there are no memtables, as I realized that the node coordinating the repair may have missed some writes, so it's best to just do a single repair just in case.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairAssignmentIterator.java:
```diff
@@ -0,0 +1,84 @@
+import java.util.NoSuchElementException;
+
+/**
+ * Convenience {@link Iterator} implementation for to assist implementations of
```
English is not my first language, but this sounds wrong, feel free to ignore if this is correct
```suggestion
 * Convenience {@link Iterator} implementation to assist implementations of
```

tolbertam replied:
You are quite right, this is grammatically incorrect! I'll fix it.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairAssignmentIterator.java:
```diff
@@ -0,0 +1,84 @@
+/**
+ * Convenience {@link Iterator} implementation for to assist implementations of
+ * {@link IAutoRepairTokenRangeSplitter#getRepairAssignments(boolean, List)} by passing {@link KeyspaceRepairPlan}
+ * to a custom {@link #nextInternal(int, KeyspaceRepairPlan)} in priority order.
```
```suggestion
 * to a custom {@link #nextInternal(int, KeyspaceRepairPlan)} method in priority order.
```

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairAssignmentIterator.java:
```diff
@@ -0,0 +1,84 @@
+    private Iterator<KeyspaceRepairPlan> currentIterator = null;
+    private PrioritizedRepairPlan currentPlan = null;
+
+    RepairAssignmentIterator(List<PrioritizedRepairPlan> repairPlans)
```
this is a public class, but the constructor is only package private, maybe this should be protected instead? so other implementations can extend without needing to hack into the `org.apache.cassandra.repair.autorepair` package.

tolbertam replied:
Ah, this is an oversight.  I wanted for this to be public so custom implementations can use this, as it cuts down some boilerplate on producing an `Iterator<KeyspaceRepairAssignments>`.  Will make this public.

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairAssignmentIterator.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public boolean hasNext()
+    {
+        return currentIterator().hasNext();
```
potential NPE here, when in the `currentIterator()` method, `currentIterator` is `null` __AND__ `!repairPlanIterator.hasNext()` (empty repair plan iterator). I recommend handling the null here:

```
```suggestion
        Iterator<KeyspaceRepairPlan> iterator = currentIterator();
        return iterator != null && iterator.hasNext();
```

frankgh commented on src/java/org/apache/cassandra/repair/autorepair/RepairAssignmentIterator.java:
```diff
@@ -0,0 +1,84 @@
+    @Override
+    public KeyspaceRepairAssignments next()
+    {
+        if (!currentIterator.hasNext())
```
also potential for an NPE, but this should probably just leverage `hasNext()`?
```suggestion
        if (!hasNext())
```

frankgh commented on src/java/org/apache/cassandra/service/AutoRepairServiceMBean.java:
```diff
@@ -71,4 +70,8 @@ public interface AutoRepairServiceMBean
 
     public Set<String> getOnGoingRepairHostIds(RepairType rType);
+
+    public Map<String, String> getAutoRepairTokenRangeSplitterParameters(RepairType repairType);
```
I don't think we can expose non-core JDK types through mBeans, we try to expose only native types . I see issues with both `RepairType` and `InetAddressAndPort` in this class. Especially concerning for `InetAddressAndPort`, we have methods to stringify the value see `org.apache.cassandra.service.StorageService#getNaturalEndpointsWithPort(java.lang.String, java.lang.String, java.lang.String)`. Maybe exposing an enum is fine though, so `RepairType` might be okay, but I have not seen this being done anywhere else in the beans that we expose through JMX

tolbertam replied:
Ah great point.  Given `RepairType` is a simple enum it might be ok, but I agree that exposing `AutoRepairConfig` and `InetAddressAndPort`d can be a problem, especially if we ever evolve the type.  I think I agree we should try to change this to only expose java types , especially since I don't see precedence for doing this anywhere else in C*.

@jaydeepkumar1984 thoughts?  Maybe we can attack this in another jira.

frankgh reviewed (commented):
I have a few more comments. Let me know if it makes sense

jaydeepkumar1984 committed:
Fix the the total expected sub-ranges calculation keeping v-nodes in mind

tolbertam committed:
Feedback and update tests

* Promote node not being present in gossip to warn
* Clean up NUMBER_OF_SUBRANGES doc
* Simplify default map parameter parsing in both splitters
* Doc cleanup and make RepairAssignmentIterator fully public
* Remove unnecessary getters and setters in Splitters
* Handle case where ColumnFamilyStore not retrievable, in this case
  return no assignments as we can assume deleted.
* Always return a RepairAssignment for a table, even if empty, in
  RepairTokenRangeSplitter as node may have missed writes
* Add set|getParameters tests
* Update AutoRepairParameterizedTest to use RepairTokenRangeSplitter
* Move no-split specific test to FixedSplitTokenRangeSplitterTest

frankgh reviewed (approved):
+1
